### PR TITLE
Add the possibility to print ol.style.Text

### PR DESCRIPTION
**Pull request by Jenselme**

### CODE REVIEW

File: src/services/print.js [hunk 1]
```diff
@@ -488,6 +492,49 @@ ngeo.Print.prototype.encodeVectorStyleStroke_ =
...
 /**
+ * @param {Array.<MapFishPrintSymbolizer>} symbolizers Array of MapFish Print
+ *     symbolizers.
+ * @param {!ol.style.Text} textStyle Text style.
+ * @private
+ */
+ngeo.Print.prototype.encodeTextStyle_ = function(symbolizers, textStyle) {
+  var symbolizer = {
+    type: 'Text'
+  };
```
Comment (elemoine) [inline src/services/print.js]: I think we need a new type for this in the mapfish-print-v3.js externs file.

File: src/services/print.js [hunk 2]
```diff
@@ -488,6 +492,49 @@ ngeo.Print.prototype.encodeVectorStyleStroke_ =
...
+  var font = textStyle.getFont().split(' ');
+  var labelAlign = textStyle.getTextAlign();
+  if (labelAlign) {
+    symbolizer.labelAlign = labelAlign;
+  }
+  var labelRotation = textStyle.getRotation();
+  if (labelRotation) {
+    symbolizer.labelRotation = labelRotation;
```
Comment (elemoine) [inline src/services/print.js]: Do we need to convert from radians to degrees here?

File: src/services/print.js (hunk 1 above)
Comment (Jenselme) [inline src/services/print.js]: I don't know. I have no particular experience with closure. If you give me a hit, I may be able to write it.

File: src/services/print.js (hunk 2 above)
Comment (Jenselme) [inline src/services/print.js]: We need, thanks for pointing that out. PR updated!

File: src/services/print.js (hunk 1 above)
Comment (elemoine) [inline src/services/print.js]: Look at the [MapFishPrintSymbolizerPolygon type](https://github.com/camptocamp/ngeo/blob/master/externs/mapfish-print-v3.js#L428-L462) as an example. You'd need something similar for Text.

Comment (Jenselme) [inline src/services/print.js]: Type added

File: externs/mapfish-print-v3.js [hunk 3]
```diff
@@ -460,3 +460,70 @@ MapFishPrintSymbolizerPolygon.prototype.strokeOpacity;
...
+ */
+MapFishPrintSymbolizerText.prototype.labelRotation;
+
+/**
+ * @type {string}
+ */
+MapFishPrintSymbolizerText.prototype.fontWeight;
+
+/**
+ * @type {string}
```
Comment (elemoine) [inline externs/mapfish-print-v3.js]: `fontSize` is expected to be a `string` too?

File: src/services/print.js [hunk 4]
```diff
@@ -374,6 +374,10 @@ ngeo.Print.prototype.encodeVectorStyle_ =
   } else if (!goog.isNull(imageStyle)) {
     this.encodeVectorStylePoint_(styleObject.symbolizers, imageStyle);
   }
+
+  if (!goog.isNull(textStyle) && textStyle.getText()) {
```
Comment (elemoine) [inline src/services/print.js]: Please use `goog.isDef(textStyle.getText())`.

Comment (elemoine) [inline src/services/print.js]: Or maybe remove that test and test that within the `encodeTextStyle_` function. I think that would make more sense.

File: src/services/print.js [hunk 5]
```diff
@@ -488,6 +492,63 @@ ngeo.Print.prototype.encodeVectorStyleStroke_ =
...
+ *     symbolizers.
+ * @param {!ol.style.Text} textStyle Text style.
+ * @private
+ */
+ngeo.Print.prototype.encodeTextStyle_ = function(symbolizers, textStyle) {
+  var symbolizer = /** @type {MapFishPrintSymbolizerText} */ ({
+    type: 'Text'
+  });
+  var label = textStyle.getText();
+  if (label) {
```
Comment (elemoine) [inline src/services/print.js]: Do you want to use `goog.isDef(label)` here?

File: src/services/print.js [hunk 6]
```diff
@@ -488,6 +492,63 @@ ngeo.Print.prototype.encodeVectorStyleStroke_ =
...
+    symbolizer.label = label;
+  }
+
+  var labelAlign = textStyle.getTextAlign();
+  if (labelAlign) {
```
Comment (elemoine) [inline src/services/print.js]: Please consider using `goog.isDef(labelAlign)`?

File: src/services/print.js [hunk 7]
```diff
@@ -488,6 +492,63 @@ ngeo.Print.prototype.encodeVectorStyleStroke_ =
...
+    symbolizer.labelAlign = labelAlign;
+  }
+
+  var labelRotation = textStyle.getRotation();
+  if (labelRotation) {
```
Comment (elemoine) [inline src/services/print.js]: `goog.isDef`

File: src/services/print.js [hunk 8]
```diff
@@ -488,6 +492,63 @@ ngeo.Print.prototype.encodeVectorStyleStroke_ =
...
+    // Mapfish print expects a string, not a number to rotate text
+    symbolizer.labelRotation = (labelRotation * 180 / Math.PI).toString();
+  }
+
+  var font = textStyle.getFont().split(' ');
```
Comment (elemoine) [inline src/services/print.js]: `getFont` may return an `undefined` value.

File: src/services/print.js [hunk 9]
```diff
@@ -488,6 +492,63 @@ ngeo.Print.prototype.encodeVectorStyleStroke_ =
...
+    symbolizer.fontFamily = font.splice(2).join(' ');
+  }
+
+  var strokeStyle = textStyle.getStroke();
+  if (!goog.isNull(strokeStyle)) {
+    var strokeColorRgba = ol.color.asArray(strokeStyle.getColor());
+    symbolizer.haloColor = goog.color.rgbArrayToHex(strokeColorRgba);
+    symbolizer.haloOpacity = strokeColorRgba[3];
+    var width = strokeStyle.getWidth();
+    if (width) {
```
Comment (elemoine) [inline src/services/print.js]: `goog.isDef`?

**Comment by elemoine** [comment]
I've just added a few comments. But it looks good. I'll merge your branch when you've addressed the comments. Thanks!

File: src/services/print.js [hunk 10]
```diff
@@ -488,6 +492,63 @@ ngeo.Print.prototype.encodeVectorStyleStroke_ =
 
 
 /**
+ * @param {Array.<MapFishPrintSymbolizer>} symbolizers Array of MapFish Print
+ *     symbolizers.
+ * @param {!ol.style.Text} textStyle Text style.
+ * @private
+ */
+ngeo.Print.prototype.encodeTextStyle_ = function(symbolizers, textStyle) {
+  var symbolizer = /** @type {MapFishPrintSymbolizerText} */ ({
```
Comment (gberaudo) [inline src/services/print.js]: `MapFishPrintSymbolizerText` is a constructor, is it fine to cast a hash to an instance?
Have you considered instanciating the class?

Comment (elemoine) [inline src/services/print.js]: It's not really a constructor. It is a constructor because it is the way types are defined in an externs file.

Comment (gberaudo) [inline src/services/print.js]: You are right.

File: externs/mapfish-print-v3.js (hunk 3 above)
Comment (Jenselme) [inline externs/mapfish-print-v3.js]: Yes. By instance, `fontSize` is `16px`.

**Comment by Jenselme** [comment]
If have just updated the PR to address all your comments.

File: externs/mapfish-print-v3.js (hunk 3 above)
Comment (elemoine) [inline externs/mapfish-print-v3.js]: Obviously.

**Pull request merged**